processor/transformprocessor/internal/common/traces.go

```diff
@@ -0,0 +1,137 @@
+	"context"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/multierr"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlresource"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlscope"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlspanevent"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottltraces"
+)
+
+type TracesContext interface {
+	ProcessTraces(ctx context.Context, td ptrace.Traces) error
+}
+
+var _ TracesContext = &traceStatements{}
+
+type traceStatements []*ottl.Statement[ottltraces.TransformContext]
+
+func (t traceStatements) ProcessTraces(ctx context.Context, td ptrace.Traces) error {
+	for i := 0; i < td.ResourceSpans().Len(); i++ {
+		rspans := td.ResourceSpans().At(i)
+		for j := 0; j < rspans.ScopeSpans().Len(); j++ {
+			sspans := rspans.ScopeSpans().At(j)
+			spans := sspans.Spans()
+			for k := 0; k < spans.Len(); k++ {
+				tCtx := ottltraces.NewTransformContext(spans.At(k), sspans.Scope(), rspans.Resource())
+				for _, statement := range t {
+					_, _, err := statement.Execute(ctx, tCtx)
+					if err != nil {
+						return err
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+var _ TracesContext = &spanEventStatements{}
+
+type spanEventStatements []*ottl.Statement[ottlspanevent.TransformContext]
+
+func (s spanEventStatements) ProcessTraces(ctx context.Context, td ptrace.Traces) error {
+	for i := 0; i < td.ResourceSpans().Len(); i++ {
+		rspans := td.ResourceSpans().At(i)
+		for j := 0; j < rspans.ScopeSpans().Len(); j++ {
+			sspans := rspans.ScopeSpans().At(j)
+			spans := sspans.Spans()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+				spanEvents := span.Events()
+				for n := 0; n < spanEvents.Len(); n++ {
+					tCtx := ottlspanevent.NewTransformContext(spanEvents.At(k), span, sspans.Scope(), rspans.Resource())
+					for _, statement := range s {
+						_, _, err := statement.Execute(ctx, tCtx)
+						if err != nil {
+							return err
+						}
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+type TraceParserCollection struct {
+	parserCollection
+	traceParser     ottl.Parser[ottltraces.TransformContext]
+	spanEventParser ottl.Parser[ottlspanevent.TransformContext]
+}
+
+func NewTraceParserCollection(functions map[string]interface{}, settings component.TelemetrySettings) TraceParserCollection {
+	return TraceParserCollection{
+		parserCollection: parserCollection{
+			settings:       settings,
+			resourceParser: ottlresource.NewParser(ResourceFunctions(), settings),
+			scopeParser:    ottlscope.NewParser(ScopeFunctions(), settings),
+		},
+		traceParser: ottltraces.NewParser(functions, settings),
+	}
+}
+
+func (pc TraceParserCollection) ParseContextStatements(contextStatements []ContextStatements) ([]TracesContext, error) {
```

Comment:
What is the reason to return a slice? I think we can return just one and call the statements sequentially.

Reply:
We could, but it will mean implementing loops in `config.Validate` and each processor's signal's NewProcessor function

Reply:
I think otherwise you will duplicate the loops in the user code (consumer of the package).

Reply:
@bogdandrutu If we remove this function's ability to handle a list of ContextStatements then the NewProcessor and Validate functions all look like:

```go
type Processor struct {
	contexts []consumer.Logs
}

func NewProcessor(contextStatements []common.ContextStatements, settings component.TelemetrySettings) (*Processor, error) {
	pc, err := common.NewLogParserCollection(Functions(), settings)
	if err != nil {
		return nil, err
	}
	contexts := make([]consumer.Logs, len(contextStatements))
	for i, c := range contextStatements {
		context, err := pc.ParseContextStatements(c)
		if err != nil {
			return nil, err
		}
		contexts[i] = context
	}
	return &Processor{
		contexts: contexts,
	}, nil
}
```
I would definitely prefer to hide this shared logic in the `ParseContextStatement` functions.  In the future, if this all moves into the OTTL, we can also expose a function for handling single `ContextStatements` if necessary